FastSimulation/Tracking/plugins/TrajectorySeedProducer.cc

```diff
@@ -173,51 +188,66 @@ void TrajectorySeedProducer::produce(edm::Event& e, const edm::EventSetup& es)
     TrackingRegion * selectedTrackingRegion = 0;
+    auto pixelTripletGeneratorPtr = pixelTripletGenerator.get();
+    auto MultiHitGeneratorPtr = MultiHitGenerator.get();
+    if(MultiHitGenerator)
+    {
+	MultiHitGenerator->initES(es);
+    }
+    std::unique_ptr<HitDoublets> hitDoublets;
        
     // define a lambda function
     // to select hit pairs, triplets, ... compatible with the region
-    SeedFinder::Selector selectorFunction = [&es,&measurementTracker,&selectedTrackingRegion](const std::vector<const FastTrackerRecHit*>& hits) -> bool
+    SeedFinder::Selector selectorFunction = [&es,&measurementTracker,&selectedTrackingRegion,&pixelTripletGeneratorPtr,&MultiHitGeneratorPtr,&hitDoublets](const std::vector<const FastTrackerRecHit*>& hits) mutable -> bool
     {
 	// criteria for hit pairs
 	// based on HitPairGeneratorFromLayerPair::doublets( const TrackingRegion& region, const edm::Event & iEvent, const edm::EventSetup& iSetup, Layers layers)
-	if (hits.size()==2)
+	if(hits.size()==2)
 	{
-	    const FastTrackerRecHit * innerHit = hits[0];
-	    const FastTrackerRecHit * outerHit = hits[1];
+	    const FastTrackerRecHit * firstHit = hits[0];
+	    const FastTrackerRecHit * secondHit = hits[1];
 	    
-	    const DetLayer * innerLayer = measurementTracker->geometricSearchTracker()->detLayer(innerHit->det()->geographicalId());
-	    const DetLayer * outerLayer = measurementTracker->geometricSearchTracker()->detLayer(outerHit->det()->geographicalId());
+	    const DetLayer * firstLayer = measurementTracker->geometricSearchTracker()->detLayer(firstHit->det()->geographicalId());
+	    const DetLayer * secondLayer = measurementTracker->geometricSearchTracker()->detLayer(secondHit->det()->geographicalId());
 	    
-	    std::vector<BaseTrackerRecHit const *> innerHits(1,(const BaseTrackerRecHit*) innerHit->hit());
-	    std::vector<BaseTrackerRecHit const *> outerHits(1,(const BaseTrackerRecHit*) outerHit->hit());
-		
-	    const RecHitsSortedInPhi* ihm=new RecHitsSortedInPhi (innerHits, selectedTrackingRegion->origin(), innerLayer);
-	    const RecHitsSortedInPhi* ohm=new RecHitsSortedInPhi (outerHits, selectedTrackingRegion->origin(), outerLayer);
-		
-	    HitDoublets result(*ihm,*ohm); 
-	    HitPairGeneratorFromLayerPair::doublets(*selectedTrackingRegion,*innerLayer,*outerLayer,*ihm,*ohm,es,0,result);
-		
-	    if(result.size()!=0)
-	    {
-		return true; 
-	    }
-	    else
-	    {
-		return false;
-	    }
+	    std::vector<BaseTrackerRecHit const *> firstHits(1,(const BaseTrackerRecHit*) firstHit->hit());
+	    std::vector<BaseTrackerRecHit const *> secondHits(1,(const BaseTrackerRecHit*) secondHit->hit());
+
+	    const RecHitsSortedInPhi* fhm=new RecHitsSortedInPhi (firstHits, selectedTrackingRegion->origin(), firstLayer);
+	    const RecHitsSortedInPhi* shm=new RecHitsSortedInPhi (secondHits, selectedTrackingRegion->origin(), secondLayer);
+	    hitDoublets.reset(new HitDoublets(*fhm,*shm));
```

Comment:
Here you store data in a pointer outside the lambda. This may cause problems because it depends on the hits when this method is called.